app/controllers/concerns/parameters.rb

```diff
@@ -5,59 +5,67 @@ module Parameters
   extend ActiveSupport::Concern
 
   ParamType = ActionController::Parameters
-
-  RELATIONSHIP_TYPE = ParamType.map(
-    id: ParamType.string,
-    type: ParamType.string
+  ParamType.action_on_unpermitted_parameters = :raise
+  ID_TYPE = ParamType.integer | ParamType.string
+  ARRAY_OR_STRING = ParamType.array(ParamType.string) | ParamType.string
+  POSITIVE_NUMBER = ParamType.integer & ParamType.gt(0)
+  DEFAULT_PERMITTED = StrongerParameters::ControllerSupport::PermittedParameters::DEFAULT_PERMITTED.merge(
+    _json: ParamType.nil,
+    include: ParamType.regexp(/^[a-zA-Z0-9,_]*$/)
   )
 
-  included do
-    def relationships_enabled?
-      params.permit(relationships: ParamType.boolean)
-            .with_defaults(relationships: true)[:relationships]
-    end
+  class_methods do
+    attr_accessor :__permitted_params_for_action
 
-    def resource_params
-      params.permit(data: ParamType.map(
-        attributes: ParamType.map,
-        relationships: ParamType.map
-      ))
-      params.require(:data).permit(attributes: {}, relationships: {})
+    def permitted_params_for_action(action, params)
+      self.__permitted_params_for_action ||= {}
+      self.__permitted_params_for_action[action] = params
     end
+  end
 
-    def resource_attributes
-      resource_params[:attributes]&.permit(
-        serializer.attributes_to_serialize.keys
-      )
-    end
+  included do
+    permitted_params_for_action :index, {
```

Comment:
Would it be possible to make these default and when concrete module calls `permitted_params_for_action` it would append new ones. Of course by having an option to fully override?

Reply:
These are the default ones :see_no_evil: for any index route, and each controller can append to these. So I don't really understand what you're proposing :thinking:

Reply:
I see, there are `parent_params` that are being merged.